// @mcp-use/inspector/dist/server/chunk-S2MXUMF5.js
import {
  getInspectorVersion
} from "./chunk-4CQ52PTY.js";
import {
  checkClientFiles,
  getClientDistPath,
  getContentType
} from "./chunk-IDMHJUEM.js";

// src/server/shared-static.ts
import { existsSync, readFileSync } from "fs";
import { join } from "path";
var USE_CDN = process.env.INSPECTOR_USE_CDN === "true";
var INSPECTOR_VERSION = getInspectorVersion();
var CDN_BASE = process.env.INSPECTOR_CDN_BASE ?? "https://inspector-cdn.mcp-use.com";
var CDN_JS_URL = `${CDN_BASE}/inspector@${INSPECTOR_VERSION}.js`;
var CDN_CSS_URL = `${CDN_BASE}/inspector@${INSPECTOR_VERSION}.css`;
function injectRuntimeConfig(html, config) {
  if (!config) return html;
  const scripts = [];
  if (config.devMode) {
    scripts.push(`<script>window.__MCP_DEV_MODE__ = true;</script>`);
  }
  if (config.sandboxOrigin) {
    scripts.push(
      `<script>window.__MCP_SANDBOX_ORIGIN__ = ${JSON.stringify(config.sandboxOrigin)};</script>`
    );
  }
  if (scripts.length === 0) return html;
  const injection = scripts.join("\n    ");
  return html.replace("</head>", `    ${injection}
  </head>`);
}
function generateCdnShellHtml(config) {
  const runtimeScripts = (() => {
    if (!config) return "";
    const scripts = [];
    if (config.devMode) {
      scripts.push(`<script>window.__MCP_DEV_MODE__ = true;</script>`);
    }
    if (config.sandboxOrigin) {
      scripts.push(
        `<script>window.__MCP_SANDBOX_ORIGIN__ = ${JSON.stringify(config.sandboxOrigin)};</script>`
      );
    }
    return scripts.join("\n    ");
  })();
  return `<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link
      rel="icon"
      type="image/svg+xml"
      href="${CDN_BASE}/favicon-black.svg"
    />
    <link
      rel="icon"
      type="image/svg+xml"
      href="${CDN_BASE}/favicon-white.svg"
      media="(prefers-color-scheme: dark)"
    />
    <link
      rel="icon"
      type="image/svg+xml"
      href="${CDN_BASE}/favicon-black.svg"
      media="(prefers-color-scheme: light)"
    />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link
      href="https://fonts.googleapis.com/css2?family=Ubuntu:wght@400;500;700&display=swap"
      rel="stylesheet"
    />
    <link rel="stylesheet" href="${CDN_CSS_URL}" />
    <title>Inspector | mcp-use</title>
    <script>window.__INSPECTOR_VERSION__ = ${JSON.stringify(INSPECTOR_VERSION)};</script>
    ${runtimeScripts}
  </head>
  <body>
    <script>
      if (typeof window !== "undefined" && typeof window.process === "undefined") {
        window.process = {
          env: {},
          platform: "browser",
          browser: true,
          version: "v18.0.0",
          versions: { node: "18.0.0" },
          cwd: () => "/",
          nextTick: (fn, ...args) => queueMicrotask(() => fn(...args)),
        };
      }
    </script>
    <div id="root"></div>
    <script type="module" src="${CDN_JS_URL}"></script>
  </body>
</html>`;
}
function registerStaticRoutes(app, clientDistPath, runtimeConfig) {
  if (USE_CDN) {
    const serveShell = (c) => c.html(generateCdnShellHtml(runtimeConfig));
    app.get("/", (c) => {
      const url = new URL(c.req.url);
      return c.redirect(`/inspector${url.search}`);
    });
    app.get("/inspector", serveShell);
    app.get("/inspector/*", serveShell);
    app.post("/inspector/*", serveShell);
    app.get("*", serveShell);
    return;
  }
  const distPath = clientDistPath || getClientDistPath();
  if (!checkClientFiles(distPath)) {
    console.warn(`\u26A0\uFE0F  MCP Inspector client files not found at ${distPath}`);
    console.warn(
      `   Run 'yarn build' in the inspector package to build the UI`
    );
    app.get("*", (c) => {
      return c.html(`
      <!DOCTYPE html>
      <html>
        <head>
          <title>MCP Inspector</title>
        </head>
        <body>
          <h1>MCP Inspector</h1>
          <p>Client files not found. Please run 'yarn build' to build the UI.</p>
        </body>
      </html>
    `);
    });
    return;
  }
  app.get("/inspector/assets/*", (c) => {
    const path = c.req.path.replace("/inspector/assets/", "assets/");
    const fullPath = join(distPath, path);
    if (existsSync(fullPath)) {
      const content = readFileSync(fullPath);
      const contentType = getContentType(fullPath);
      c.header("Content-Type", contentType);
      return c.body(content);
    }
    return c.notFound();
  });
  app.get("/", (c) => {
    const url = new URL(c.req.url);
    const queryString = url.search;
    return c.redirect(`/inspector${queryString}`);
  });
  const serveIndex = (c) => {
    const indexPath = join(distPath, "index.html");
    if (existsSync(indexPath)) {
      const content = injectRuntimeConfig(
        readFileSync(indexPath, "utf-8"),
        runtimeConfig
      );
      return c.html(content);
    }
    return c.html(`
      <!DOCTYPE html>
      <html>
        <head>
          <title>MCP Inspector</title>
        </head>
        <body>
          <h1>MCP Inspector</h1>
          <p>Client files not found. Please run 'yarn build' to build the UI.</p>
        </body>
      </html>
    `);
  };
  app.get("/inspector", serveIndex);
  app.get("/inspector/*", serveIndex);
  app.post("/inspector/*", serveIndex);
  app.get("*", (c) => {
    const indexPath = join(distPath, "index.html");
    if (existsSync(indexPath)) {
      const content = injectRuntimeConfig(
        readFileSync(indexPath, "utf-8"),
        runtimeConfig
      );
      return c.html(content);
    }
    return c.html(`
      <!DOCTYPE html>
      <html>
        <head>
          <title>MCP Inspector</title>
        </head>
        <body>
          <h1>MCP Inspector</h1>
          <p>Client files not found. Please run 'yarn build' to build the UI.</p>
        </body>
      </html>
    `);
  });
}
function registerStaticRoutesWithDevProxy(app, clientDistPath) {
  const distPath = clientDistPath || getClientDistPath();
  const isDev = process.env.NODE_ENV === "development" || process.env.VITE_DEV === "true";
  if (isDev) {
    console.warn(
      "\u{1F527} Development mode: Proxying client requests to Vite dev server"
    );
    app.get("*", async (c) => {
      const path = c.req.path;
      if (path.startsWith("/api/") || path.startsWith("/inspector/api/") || path === "/inspector/config.json") {
        return c.notFound();
      }
      try {
        const viteUrl = `http://localhost:3000${path}`;
        const response = await fetch(viteUrl, {
          signal: AbortSignal.timeout(1e3)
        });
        if (response.ok) {
          const content = await response.text();
          const contentType = response.headers.get("content-type") || "text/html";
          c.header("Content-Type", contentType);
          return c.html(content);
        }
      } catch (error) {
        console.warn(`Failed to proxy to Vite dev server: ${error}`);
      }
      return c.html(`
        <!DOCTYPE html>
        <html>
          <head>
            <title>MCP Inspector - Development</title>
          </head>
          <body>
            <h1>MCP Inspector - Development Mode</h1>
            <p>Vite dev server is not running. Please start it with:</p>
            <pre>yarn dev:client</pre>
            </body>
        </html>
      `);
    });
  } else {
    registerStaticRoutes(app, distPath);
  }
}

export {
  registerStaticRoutes,
  registerStaticRoutesWithDevProxy
};
