// @mcp-use/inspector/dist/server/chunk-CVECQ7BJ.js
import {
  __publicField
} from "./chunk-PKBMQBKP.js";

// src/server/rpc-log-bus.ts
var SimpleEventEmitter = class {
  constructor() {
    __publicField(this, "listeners", /* @__PURE__ */ new Map());
  }
  on(event, listener) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, /* @__PURE__ */ new Set());
    }
    this.listeners.get(event).add(listener);
  }
  off(event, listener) {
    this.listeners.get(event)?.delete(listener);
  }
  emit(event, ...args) {
    this.listeners.get(event)?.forEach((listener) => {
      try {
        listener(...args);
      } catch (e) {
        console.error("Error in event listener:", e);
      }
    });
  }
};
var RpcLogBus = class {
  constructor() {
    __publicField(this, "emitter", new SimpleEventEmitter());
    __publicField(this, "bufferByServer", /* @__PURE__ */ new Map());
  }
  publish(event) {
    const buffer = this.bufferByServer.get(event.serverId) ?? [];
    buffer.push(event);
    if (buffer.length > 1e3) {
      buffer.shift();
    }
    this.bufferByServer.set(event.serverId, buffer);
    this.emitter.emit("event", event);
  }
  subscribe(serverIds, listener) {
    const filter = new Set(serverIds);
    const handler = (event) => {
      if (filter.size === 0 || filter.has(event.serverId)) listener(event);
    };
    this.emitter.on("event", handler);
    return () => this.emitter.off("event", handler);
  }
  getBuffer(serverIds, limit) {
    const filter = new Set(serverIds);
    const all = [];
    for (const [serverId, buf] of this.bufferByServer.entries()) {
      if (filter.size > 0 && !filter.has(serverId)) continue;
      all.push(...buf);
    }
    all.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
    if (limit === 0) return [];
    if (!Number.isFinite(limit) || limit < 0) return all;
    return all.slice(0, limit);
  }
  clear(serverIds) {
    if (serverIds && serverIds.length > 0) {
      const filter = new Set(serverIds);
      for (const serverId of filter) {
        this.bufferByServer.delete(serverId);
      }
    } else {
      this.bufferByServer.clear();
    }
  }
};
var rpcLogBus = new RpcLogBus();

export {
  rpcLogBus
};
