// @mcp-use/inspector/dist/server/chunk-42S46FP7.js
import {
  registerInspectorRoutes
} from "./chunk-JNAYXSVD.js";
import {
  registerStaticRoutes
} from "./chunk-S2MXUMF5.js";
import {
  checkClientFiles,
  getClientDistPath
} from "./chunk-IDMHJUEM.js";

// src/server/middleware.ts
import { Hono } from "hono";
function mountInspector(app, config) {
  const clientDistPath = getClientDistPath();
  if (!checkClientFiles(clientDistPath)) {
    console.warn(
      `\u26A0\uFE0F  MCP Inspector client files not found at ${clientDistPath}`
    );
    console.warn(
      `   Run 'yarn build' in the inspector package to build the UI`
    );
  }
  const runtimeConfig = {
    devMode: config?.devMode,
    sandboxOrigin: config?.sandboxOrigin
  };
  if (app instanceof Hono) {
    registerInspectorRoutes(app, config);
    registerStaticRoutes(app, clientDistPath, runtimeConfig);
    return;
  }
  const honoApp = new Hono();
  registerInspectorRoutes(honoApp, config);
  registerStaticRoutes(honoApp, clientDistPath, runtimeConfig);
  app.use((req, res, next) => {
    const url = new URL(req.url || "", `http://${req.headers.host}`);
    const request = new Request(url, {
      method: req.method,
      headers: req.headers,
      body: req.method !== "GET" && req.method !== "HEAD" ? JSON.stringify(req.body) : void 0
    });
    Promise.resolve(honoApp.fetch(request)).then(async (fetchResponse) => {
      res.status(fetchResponse.status);
      fetchResponse.headers.forEach((value, key) => {
        res.setHeader(key, value);
      });
      if (fetchResponse.body) {
        const reader = fetchResponse.body.getReader();
        while (true) {
          const { done, value } = await reader.read();
          if (done) {
            break;
          }
          res.write(value);
        }
        res.end();
      } else {
        res.end();
      }
    }).catch(next);
  });
}

export {
  mountInspector
};
