// @mcp-use/inspector/dist/server/chunk-O7EUTZ5E.js
// src/server/utils.ts
function isValidUrl(urlString) {
  try {
    const url = new URL(urlString);
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}
async function findAvailablePort(startPort = 8080, maxAttempts = 100) {
  const net = await import("net");
  for (let port = startPort; port < startPort + maxAttempts; port++) {
    try {
      await new Promise((resolve, reject) => {
        const server = net.createServer();
        server.listen(port, () => {
          server.close(() => resolve());
        });
        server.on("error", (err) => reject(err));
      });
      return port;
    } catch {
      continue;
    }
  }
  throw new Error(
    `No available port found after trying ${maxAttempts} ports starting from ${startPort}`
  );
}
async function isPortAvailable(port) {
  const net = await import("net");
  return new Promise((resolve) => {
    const server = net.createServer();
    server.listen(port, () => {
      server.close(() => resolve(true));
    });
    server.on("error", () => resolve(false));
  });
}
function parsePortFromArgs() {
  const portArgIndex = process.argv.indexOf("--port");
  if (portArgIndex !== -1 && portArgIndex + 1 < process.argv.length) {
    const portValue = Number.parseInt(process.argv[portArgIndex + 1], 10);
    if (!Number.isNaN(portValue) && portValue >= 1 && portValue <= 65535) {
      return portValue;
    }
  }
  return null;
}
function formatErrorResponse(error, context) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const errorMessage = error instanceof Error ? error.message : "Unknown error";
  const errorStack = error instanceof Error ? error.stack : void 0;
  console.error(`[${timestamp}] Error in ${context}:`, {
    message: errorMessage,
    stack: errorStack
  });
  return {
    error: errorMessage,
    context,
    timestamp,
    // Only include stack in development mode
    ...process.env.NODE_ENV === "development" && errorStack ? { stack: errorStack } : {}
  };
}

export {
  isValidUrl,
  findAvailablePort,
  isPortAvailable,
  parsePortFromArgs,
  formatErrorResponse
};
