// @mcp-use/inspector/dist/server/transport-wrapper.js
import {
  rpcLogBus
} from "./chunk-CVECQ7BJ.js";
import {
  __publicField
} from "./chunk-PKBMQBKP.js";

// src/server/transport-wrapper.ts
function wrapTransportForLogging(transport, serverId) {
  class LoggingTransport {
    constructor(inner) {
      this.inner = inner;
      __publicField(this, "onclose");
      __publicField(this, "onerror");
      __publicField(this, "onmessage");
      this.inner.onmessage = (message, extra) => {
        try {
          rpcLogBus.publish({
            serverId,
            direction: "receive",
            timestamp: (/* @__PURE__ */ new Date()).toISOString(),
            message
          });
        } catch {
        }
        this.onmessage?.(message, extra);
      };
      this.inner.onclose = () => {
        this.onclose?.();
      };
      this.inner.onerror = (error) => {
        this.onerror?.(error);
      };
    }
    async start() {
      if (typeof this.inner.start === "function") {
        await this.inner.start();
      }
    }
    async send(message, options) {
      try {
        rpcLogBus.publish({
          serverId,
          direction: "send",
          timestamp: (/* @__PURE__ */ new Date()).toISOString(),
          message
        });
      } catch {
      }
      await this.inner.send(message, options);
    }
    async close() {
      await this.inner.close();
    }
    get sessionId() {
      return this.inner.sessionId;
    }
    setProtocolVersion(version) {
      if (typeof this.inner.setProtocolVersion === "function") {
        this.inner.setProtocolVersion(version);
      }
    }
  }
  return new LoggingTransport(transport);
}
export {
  wrapTransportForLogging
};
