// mcp-use/dist/chunk-MTHLLDCX.js
import {
  __name
} from "./chunk-3GQAWCBQ.js";

// src/server/utils/runtime.ts
var isDeno = typeof globalThis.Deno !== "undefined";
function getEnv(key) {
  if (isDeno) {
    return globalThis.Deno.env.get(key);
  }
  return process.env[key];
}
__name(getEnv, "getEnv");
function getCwd() {
  if (isDeno) {
    return globalThis.Deno.cwd();
  }
  return process.cwd();
}
__name(getCwd, "getCwd");
var fsHelpers = {
  async readFileSync(path, encoding = "utf8") {
    if (isDeno) {
      return await globalThis.Deno.readTextFile(path);
    }
    const { readFileSync } = await import("fs");
    const result = readFileSync(path, encoding);
    return typeof result === "string" ? result : result.toString(encoding);
  },
  async readFile(path) {
    if (isDeno) {
      const data = await globalThis.Deno.readFile(path);
      return data.buffer;
    }
    const { readFileSync } = await import("fs");
    const buffer = readFileSync(path);
    return buffer.buffer.slice(
      buffer.byteOffset,
      buffer.byteOffset + buffer.byteLength
    );
  },
  async existsSync(path) {
    if (isDeno) {
      try {
        await globalThis.Deno.stat(path);
        return true;
      } catch {
        return false;
      }
    }
    const { existsSync } = await import("fs");
    return existsSync(path);
  },
  async readdirSync(path) {
    if (isDeno) {
      const entries = [];
      for await (const entry of globalThis.Deno.readDir(path)) {
        entries.push(entry.name);
      }
      return entries;
    }
    const { readdirSync } = await import("fs");
    return readdirSync(path);
  }
};
var pathHelpers = {
  join(...paths) {
    if (isDeno) {
      return paths.join("/").replace(/\/+/g, "/");
    }
    return paths.join("/").replace(/\/+/g, "/");
  },
  relative(from, to) {
    const fromParts = from.split("/").filter((p) => p);
    const toParts = to.split("/").filter((p) => p);
    let i = 0;
    while (i < fromParts.length && i < toParts.length && fromParts[i] === toParts[i]) {
      i++;
    }
    const upCount = fromParts.length - i;
    const relativeParts = [...Array(upCount).fill(".."), ...toParts.slice(i)];
    return relativeParts.join("/");
  }
};
function generateUUID() {
  return globalThis.crypto.randomUUID();
}
__name(generateUUID, "generateUUID");

export {
  isDeno,
  getEnv,
  getCwd,
  fsHelpers,
  pathHelpers,
  generateUUID
};
