// mcp-use/dist/chunk-KUEVOU4M.js
import {
  __name
} from "./chunk-3GQAWCBQ.js";

// src/errors.ts
var ElicitationValidationError = class _ElicitationValidationError extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "ElicitationValidationError";
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, _ElicitationValidationError);
    }
  }
  static {
    __name(this, "ElicitationValidationError");
  }
};
var ElicitationTimeoutError = class _ElicitationTimeoutError extends Error {
  constructor(message, timeoutMs) {
    super(message);
    this.timeoutMs = timeoutMs;
    this.name = "ElicitationTimeoutError";
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, _ElicitationTimeoutError);
    }
  }
  static {
    __name(this, "ElicitationTimeoutError");
  }
};
var ElicitationDeclinedError = class _ElicitationDeclinedError extends Error {
  static {
    __name(this, "ElicitationDeclinedError");
  }
  constructor(message = "User declined the elicitation request") {
    super(message);
    this.name = "ElicitationDeclinedError";
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, _ElicitationDeclinedError);
    }
  }
};

export {
  ElicitationValidationError,
  ElicitationTimeoutError,
  ElicitationDeclinedError
};
